File: Trees/Python/trie_demo.py
from __future__ import annotations

import sys
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict

TEST_CASE_FILE = Path("..") / "test_case" / "trie_queries.txt"


@dataclass
class TrieNode:
    children: Dict[str, "TrieNode"] = field(default_factory=dict)
    is_end_of_word: bool = False


class Trie:
    def __init__(self) -> None:
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for char in word:
            node = node.children.setdefault(char, TrieNode())
        node.is_end_of_word = True

    def search(self, word: str) -> bool:
        node = self._find_node(word)
        return node is not None and node.is_end_of_word

    def starts_with(self, prefix: str) -> bool:
        return self._find_node(prefix) is not None

    def delete(self, word: str) -> bool:
        return self._delete(self.root, word, 0)

    def count_words_with_prefix(self, prefix: str) -> int:
        node = self._find_node(prefix)
        if node is None:
            return 0
        return self._count_words(node)

    def autocomplete(self, prefix: str) -> list[str]:
        node = self._find_node(prefix)
        if node is None:
            return []
        results: list[str] = []
        self._collect_words(node, prefix, results)
        return results

    def _find_node(self, text: str) -> TrieNode | None:
        node = self.root
        for char in text:
            if char not in node.children:
                return None
            node = node.children[char]
        return node

    def _delete(self, node: TrieNode, word: str, depth: int) -> bool:
        if depth == len(word):
            if not node.is_end_of_word:
                return False
            node.is_end_of_word = False
            return len(node.children) == 0

        char = word[depth]
        child = node.children.get(char)
        if child is None:
            return False

        should_delete_child = self._delete(child, word, depth + 1)
        if should_delete_child:
            del node.children[char]
            return not node.is_end_of_word and len(node.children) == 0
        return False

    def _count_words(self, node: TrieNode) -> int:
        total = 1 if node.is_end_of_word else 0
        for child in node.children.values():
            total += self._count_words(child)
        return total

    def _collect_words(self, node: TrieNode, prefix: str, results: list[str]) -> None:
        if node.is_end_of_word:
            results.append(prefix)
        for char in sorted(node.children.keys()):
            self._collect_words(node.children[char], prefix + char, results)


def run_queries(file_path: Path) -> int:
    if not file_path.is_file():
        print(f"Query file not found: {file_path}", file=sys.stderr)
        return 1

    trie = Trie()
    print("Trie / Prefix Tree Demo (Python)")
    print("================================")

    for raw_line in file_path.read_text().splitlines():
        line = raw_line.strip()
        if not line or line.startswith("#"):
            continue

        parts = line.split(maxsplit=1)
        command = parts[0].lower()
        argument = parts[1] if len(parts) > 1 else ""

        if command == "insert":
            trie.insert(argument)
            print(f"insert {argument} -> OK")
        elif command == "search":
            print(f"search {argument} -> {'FOUND' if trie.search(argument) else 'NOT FOUND'}")
        elif command == "prefix":
            print(f"prefix {argument} -> {'YES' if trie.starts_with(argument) else 'NO'}")
        elif command == "delete":
            existed = trie.search(argument)
            trie.delete(argument)
            print(f"delete {argument} -> {'DELETED' if existed else 'NOT FOUND'}")
        elif command == "autocomplete":
            matches = trie.autocomplete(argument)
            print(f"autocomplete {argument} -> {', '.join(matches) if matches else '(none)'}")
        elif command == "countprefix":
            print(f"countprefix {argument} -> {trie.count_words_with_prefix(argument)}")
        else:
            print(f"unknown command: {line}")

    return 0


def main() -> int:
    query_file = TEST_CASE_FILE
    if len(sys.argv) > 1:
        query_file = Path(sys.argv[1])
    return run_queries(query_file)


if __name__ == "__main__":
    raise SystemExit(main())

File: Trees/Java/src/TrieDemo.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class TrieDemo {
    private static final Path DEFAULT_QUERY_FILE = Paths.get("..", "test_case", "trie_queries.txt");

    private static class TrieNode {
        Map<Character, TrieNode> children = new TreeMap<>();
        boolean isEndOfWord = false;
    }

    private static class Trie {
        private final TrieNode root = new TrieNode();

        void insert(String word) {
            TrieNode current = root;
            for (char rawChar : word.toCharArray()) {
                char ch = Character.toLowerCase(rawChar);
                if (!Character.isLetter(ch)) {
                    continue;
                }
                current = current.children.computeIfAbsent(ch, key -> new TrieNode());
            }
            current.isEndOfWord = true;
        }

        boolean search(String word) {
            TrieNode node = findNode(word);
            return node != null && node.isEndOfWord;
        }

        boolean startsWith(String prefix) {
            return findNode(prefix) != null;
        }

        boolean delete(String word) {
            return deleteRecursive(root, normalize(word), 0).deleted;
        }

        int countPrefix(String prefix) {
            TrieNode node = findNode(prefix);
            if (node == null) {
                return 0;
            }
            return countWords(node);
        }

        List<String> autocomplete(String prefix) {
            TrieNode node = findNode(prefix);
            if (node == null) {
                return List.of();
            }
            List<String> results = new ArrayList<>();
            collectWords(node, normalize(prefix), results);
            return results;
        }

        private TrieNode findNode(String text) {
            TrieNode current = root;
            for (char ch : normalize(text).toCharArray()) {
                current = current.children.get(ch);
                if (current == null) {
                    return null;
                }
            }
            return current;
        }

        private String normalize(String text) {
            StringBuilder builder = new StringBuilder();
            for (char rawChar : text.toCharArray()) {
                char ch = Character.toLowerCase(rawChar);
                if (Character.isLetter(ch)) {
                    builder.append(ch);
                }
            }
            return builder.toString();
        }

        private DeleteResult deleteRecursive(TrieNode node, String word, int depth) {
            if (depth == word.length()) {
                if (!node.isEndOfWord) {
                    return new DeleteResult(false, false);
                }
                node.isEndOfWord = false;
                return new DeleteResult(true, node.children.isEmpty());
            }

            char ch = word.charAt(depth);
            TrieNode child = node.children.get(ch);
            if (child == null) {
                return new DeleteResult(false, false);
            }

            DeleteResult result = deleteRecursive(child, word, depth + 1);
            if (result.shouldPruneChild) {
                node.children.remove(ch);
            }

            boolean shouldPrune = !node.isEndOfWord && node.children.isEmpty();
            return new DeleteResult(result.deleted, shouldPrune);
        }

        private int countWords(TrieNode node) {
            int total = node.isEndOfWord ? 1 : 0;
            for (TrieNode child : node.children.values()) {
                total += countWords(child);
            }
            return total;
        }

        private void collectWords(TrieNode node, String prefix, List<String> results) {
            if (node.isEndOfWord) {
                results.add(prefix);
            }
            for (Map.Entry<Character, TrieNode> entry : node.children.entrySet()) {
                collectWords(entry.getValue(), prefix + entry.getKey(), results);
            }
        }

        private record DeleteResult(boolean deleted, boolean shouldPruneChild) {}
    }

    public static void main(String[] args) throws IOException {
        Path queryFile = args.length > 0 ? Paths.get(args[0]) : DEFAULT_QUERY_FILE;
        if (!Files.isRegularFile(queryFile)) {
            System.err.println("Query file not found: " + queryFile);
            return;
        }

        Trie trie = new Trie();
        System.out.println("Trie / Prefix Tree Demo (Java)");
        System.out.println("===============================");

        for (String rawLine : Files.readAllLines(queryFile)) {
            String line = rawLine.trim();
            if (line.isEmpty() || line.startsWith("#")) {
                continue;
            }

            String[] parts = line.split("\\s+", 2);
            String command = parts[0].toLowerCase();
            String argument = parts.length > 1 ? parts[1] : "";

            switch (command) {
                case "insert" -> {
                    trie.insert(argument);
                    System.out.println("insert " + argument + " -> OK");
                }
                case "search" -> System.out.println("search " + argument + " -> " + (trie.search(argument) ? "FOUND" : "NOT FOUND"));
                case "prefix" -> System.out.println("prefix " + argument + " -> " + (trie.startsWith(argument) ? "YES" : "NO"));
                case "delete" -> System.out.println("delete " + argument + " -> " + (trie.delete(argument) ? "DELETED" : "NOT FOUND"));
                case "autocomplete" -> {
                    List<String> matches = trie.autocomplete(argument);
                    System.out.println("autocomplete " + argument + " -> " + (matches.isEmpty() ? "(none)" : String.join(", ", matches)));
                }
                case "countprefix" -> System.out.println("countprefix " + argument + " -> " + trie.countPrefix(argument));
                default -> System.out.println("unknown command: " + line);
            }
        }
    }
}
